packages/eslint-plugin/src/rules/no-base-to-string.ts

```diff
@@ -202,6 +225,36 @@ export default createRule<Options, MessageIds>({
           checkExpression(expression);
         }
       },
+      ...(checkArrayJoin
+        ? {
+            'CallExpression > MemberExpression.callee > Identifier[name = "join"].property'(
+              node: TSESTree.Expression,
+            ): void {
+              const memberExpr = node.parent as TSESTree.MemberExpression;
+              const type = services.getTypeAtLocation(memberExpr.object);
+              const typeParts = tsutils.typeParts(type);
+              const isArrayOrTuple = typeParts.every(
+                part => checker.isArrayType(part) || checker.isTupleType(part),
+              );
+              if (!isArrayOrTuple) {
+                return;
+              }
+
+              for (const part of typeParts) {
+                const typeArg = checker.getTypeArguments(
```

Comment:
the argument type at index 0 has different meaning based on whether it's an array or tuple. For arrays, the 0th index apparently gives the `T` in `Array<T>`, but for tuples it seems to give the `T` in `[T, U, V]`. Consequently, the following test case is currently missed by this implementation:

```ts
["foo", {}].toString();
```

I think the logic you'd want would be more like
```ts
// array case
const typeArg = (part as ts.TypeReference).getNumberIndexType();
```
tbh the above might work for the tuple case as well (and give you back the union of the members); I'm not sure... If not, you probably want to iterate through each of the type arguments, i.e.
```ts
                checker.getTypeArguments(
                                  part as ts.TypeReference,
                ).some(/* check for toStringCertainty */);
```

Reply:
@kirkwaiblinger Thanks, I changed it to use `getNumberIndexType`, it seems to work fine with noUncheckedIndexedAccess as well. 1fafc44